src/engraving/libmscore/beam.cpp

```diff
@@ -611,1482 +506,539 @@ void Beam::layout()
+//---------------------------------------------------------
 
-            StemSlash* stemSlash = c->stemSlash();
-            if (stemSlash) {
-                stemSlash->layout();
-            }
-            Tremolo* tremolo = c->tremolo();
-            if (tremolo) {
-                tremolo->layout();
-            }
-            c->segment()->createShape(c->vStaffIdx());            // recreate shape
+void Beam::layout2(std::vector<ChordRest*> chordRests, SpannerSegmentType, int frag)
+{
+    if (chordRests.empty()) {
+        return;
+    }
+    for (auto chordRest : chordRests) {
+        if (chordRest->isChord()) {
+            toChord(chordRest)->layoutStem();
         }
     }
+    if (_distribute) {
+        // fix horizontal spacing of stems
+        LayoutBeams::respace(&chordRests);
+    }
+    Chord* startChord = toChord(chordRests[0]);
+    Chord* endChord = toChord(chordRests[chordRests.size() - 1]);
+
+    // anchor represents the middle of the beam, not the tip of the stem
+    PointF startAnchor = chordBeamAnchor(startChord);
+    PointF endAnchor = chordBeamAnchor(endChord);
+    const int startNote = _up ? startChord->upNote()->line() : startChord->downNote()->line();
+    const int endNote = _up ? endChord->upNote()->line() : endChord->downNote()->line();
+    const int interval = qAbs(startNote - endNote);
+    const bool isStartDictator = _up ? startNote < endNote : startNote > endNote;
+    const qreal quarterSpace = spatium() / 4;
+    int dictator = round((isStartDictator ? startAnchor.y() : endAnchor.y()) / quarterSpace);
+    int pointer = round((isStartDictator ? endAnchor.y() : startAnchor.y()) / quarterSpace);
+
+    const int staffLines = startChord->staff()->lines(tick());
+    const int middleLine = getMiddleStaffLine(startChord, endChord, staffLines);
+
+    int slant = computeDesiredSlant(startNote, endNote, middleLine, dictator, pointer);
+    pointer = dictator + slant;
+    bool isFlat = slant == 0;
+    bool isAscending = startNote > endNote;
+    int beamCount = getBeamCount(chordRests);
+
+    offsetBeamToRemoveCollisions(chordRests, dictator, pointer, startAnchor.x(), endAnchor.x(), isFlat, isStartDictator);
+    setValidBeamPositions(dictator, pointer, beamCount, staffLines, isStartDictator, isFlat, isAscending);
+    addMiddleLineSlant(dictator, pointer, beamCount, middleLine, interval);
+
+    startAnchor.setY(quarterSpace * (isStartDictator ? dictator : pointer));
+    endAnchor.setY(quarterSpace * (isStartDictator ? pointer : dictator));
+
+    qreal beamWidth = score()->styleMM(Sid::beamWidth).val() * mag();
+    _beamDist = beamWidth + 0.25 * spatium();
+
+    add8thSpaceSlant(isStartDictator ? startAnchor : endAnchor, dictator, pointer, beamCount, interval, middleLine, isFlat);
+
+    int fragmentIndex = (_direction == Direction::AUTO || _direction == Direction::DOWN) ? 0 : 1;
+    fragments[frag]->py1[fragmentIndex] = startAnchor.y();
+    fragments[frag]->py2[fragmentIndex] = endAnchor.y();
```

Comment:
I would think we need to check the _userModified flag at this point, and not overwrite y y coordinates for `fragments` where that's true? @Nick-Mazuk